bridge/engine/EngineFunction.h

```diff
@@ -0,0 +1,249 @@
+
+    return alignedSpace - sizeof(void*);
+}
+
+using EngineFunctionStorage = std::aligned_storage_t<calculateStorageSize(), alignof(std::max_align_t)>;
+
+class Invokable
+{
+public:
+    virtual ~Invokable() = default;
+
+private:
+    friend class ::bridge::EngineFunction;
+    virtual void invoke() const = 0;
+    virtual Invokable* moveTo(EngineFunctionStorage& newStorage) = 0;
+    virtual Invokable* copyTo(EngineFunctionStorage& newStorage) const = 0;
+};
+
+template <class Func, class... Args>
+class EngineFunctionBinder final : public Invokable
+{
+public:
+    template <class UFunc, class... UArgs>
+    EngineFunctionBinder(UFunc&& func, UArgs&&... args)
+        : mEngineFunction(std::forward<UFunc>(func)),
+          mFunctionArguments(std::forward<UArgs>(args)...)
+    {
+    }
+
+    EngineFunctionBinder(const EngineFunctionBinder& rhs)
+        : mEngineFunction(rhs.mEngineFunction),
+          mFunctionArguments(rhs.mFunctionArguments)
+    {
+    }
+
+    EngineFunctionBinder(EngineFunctionBinder&& rhs)
+        : mEngineFunction(std::move(rhs.mEngineFunction)),
+          mFunctionArguments(std::move(rhs.mFunctionArguments))
+    {
+    }
+
+    EngineFunctionBinder& operator=(const EngineFunctionBinder& rhs)
+    {
+        mEngineFunction = rhs.mEngineFunction;
+        mFunctionArguments = rhs.mFunctionArguments;
+        return *this;
+    }
+
+    EngineFunctionBinder& operator=(EngineFunctionBinder&& rhs)
+    {
+        mEngineFunction = std::move(rhs.mEngineFunction);
+        mFunctionArguments = std::move(rhs.mFunctionArguments);
+        return *this;
+    }
+
+    void operator()() const { invoke(); }
+
+private:
+    using TThis = EngineFunctionBinder<Func, Args...>;
+    friend class ::bridge::EngineFunction;
+
+    template <class T, class... UArgs>
+    void callMemberFunction(T* instance, UArgs&&... args) const
+    {
+        (instance->*mEngineFunction)(std::forward<UArgs>(args)...);
+    }
+
+    template <class TTuple, size_t... I>
+    void call(std::true_type, TTuple& t, std::index_sequence<I...>) const
+    {
+        callMemberFunction(std::get<I>(t)...);
+    }
+
+    template <class TTuple, size_t... I>
+    void call(std::false_type, TTuple& t, std::index_sequence<I...>) const
+    {
+        mEngineFunction(std::get<I>(t)...);
+    }
+
+    void invoke() const final
+    {
+        call(std::is_member_function_pointer<Func>(),
+            mFunctionArguments,
+            std::make_index_sequence<std::tuple_size<std::tuple<Args...>>::value>{});
+    }
+
+    TThis* moveTo(EngineFunctionStorage& newStorage) final { return new (&newStorage) TThis(std::move(*this)); }
+
+    TThis* copyTo(EngineFunctionStorage& newStorage) const final { return new (&newStorage) TThis(*this); }
+
+private:
+    Func mEngineFunction;
+    std::tuple<Args...> mFunctionArguments;
+};
+
+} // namespace detail
+
+class EngineFunction
```

Comment:
Engine function is a sort of of std::function with a forced inlined storage  that never allocate dynamic memory so we can use it safely on EngineThread